### PR TITLE
feat(hyd): hydraulic gear system

### LINKED ISSUE #1452: Landing Gear Gravity Extension
**Is your feature request related to a problem? Please describe.**

No, but there is a need and I would say relatively good demand for this, especially when emergency scenarios become easier to simulate in-game.

**Describe the feature you'd like to see implemented**

Implement working functionality for the landing gear gravity extension lever. This can be used when no hydraulic power is available such as the APU, and dual engine failure. This doesn't necessarily have to be fancy in terms of the logic in the short term. For now, it can simply be a call to lower the gear as if it were normally lowered. 

**References**

http://www.a320dp.com/A320_DP/landing-gear/sys-12.3.0.html

**Additional context**

Discord username (if different from GitHub):

### LINKED ISSUE #2731: [BUG] Landing Gear can be extended above 264 kts
**Mod Version**
master

**Describe the bug**

when the aircraft is above 264 knots, i was able to extend the landing gear

**To Reproduce**

1. In-flight
2. speed above 264 kts
3. select landing gear lever to "down" position

**Expected behavior**

Landing gear lever can be selected down but it will not extend the landing gear when airspeed is above 264 knots.

Landing gear can only be extended below 260 knots

**References**

A320 AMM 32-31-00-00 Normal Extension and Retraction

safety valve 3
safety valve

**Additional context**

in the A320, there is a safety valve (49GA) located in the RH main landing gear bay (inside wheel well) which isolates the Green Hydraulic System from the Landing Gear if the airspeed is above 264 knots (ADIRU 1 and ADIRU 3 as source)

The valve is always energized open (speed less than 260 kts) and de-energized to closed (speed more than 264 kts)

Was this working before/when did the issue start occurring?

no

Is this a problem in the vanilla unmodded game?

yes

Discord username (if different from GitHub): avteknisyan

### LINKED ISSUE #5874: Toggle Landing Gear keybind overrides No Gearup on Ground and allows gear lever to be moved
### Aircraft Version

Development

### Build info

```json
{
  "built": "2021-09-19T20:59:10+00:00",
  "ref": "master",
  "sha": "74045fb5eb77b3043f816d8d26028c33fb63bfd4",
  "actor": "davidwalschots",
  "event_name": "manual"
}
```

### Describe the bug

When using the mouse to move the landing gear lever up, it behaves as it should, where this is disabled on ground. 

However, using a keybind set to the ``KEY_GEAR_TOGGLE`` keyboard event allows the gear lever to be manipulated, and it ignores the default sim configuration that is supposed to manage this.

### Expected behavior

TOGGLE GEAR Keybind should not allow you to move the gear lever if you are unable to do so with the mouse.

### Steps to reproduce

1. Spawn in cold/dark or on the ground state
2. Try to move the gear lever with the mouse, the gear lever should not move
3. Use a keybind to move the gear lever (set Toggle Gear in keybinds). The gear lever moves when it should not.

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

2Cas#1022

### LINKED ISSUE #7070: Gravity gear extension
### Aircraft Version

Stable

### Description

It would be cool if we could use the gravity gear extension and perform a manual gear drop, just to add a little realism to hydraulic emergencies :)

You pull the gravity gear extension handle up, then you rotate it three times clockwise

gear extension
gearextension

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

GeeMeeKee#9761

### PR DESCRIPTION
**Pull request by crocket63**
Fixes #2731
Fixes #1452 
Fixes #5874
Fixes #7070

## Summary of Changes

This PR will provide real time physically simulated landing gear system.

- [x] Door assembly simulation
- [x] Gear assembly simulation
- [x] LGCIU state machine
- [x] Hydraulic gear system implementation
- [x] Manual extention
- [ ] Default model animation update
- [x] Inputs handling
- [x] Gear lights indicator
- [x] LGCIU coordination
- [x] Basic lgciu failures
- [x] Basic sensors failures

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Known issues:
1. Door animations are not connected to new system. DO NOT look at door animations, they are irrelevant.
Instead  check door positions on their respective simvars A32NX_GEAR_DOOR_{gear}_POSITION

2. Actuators are virtually connected to lines even when safety/cutoff valves are closed. This will result in gravity extension that can impact green circuit pressure when in use.

Normal operations:
1. Gear lever cannot be set to up position on ground whatever input event is used
2. Gear can go up
3. Gear can extend
4. Gear cannot move above 264kts
5. No LGCIU fault ever apear in nominal conditions

Abnormal operations:
1. Gear cannot move without green hydraulics
2. Gear can move if using yellow epump plus PTU with all edp off. It should be slow however, and could cause LGCIU fault.
3. LGCIU1 power failure cause an ecam warning and gear light indicator is non operative. Gear can still work normally if cycling gear lever
4. LGCIU2 power failure cause an ecam warning and gear light indicator is still operative. Gear can still work normally if cycling gear lever
5. Proximity sensor failures cause gear sequence to get stuck. Ecam fault will be trigger. After recycling gear lever gear operation should be normal.
6. Dual LGCIU power failure cause the gear to be unresponsive. Gear can be extended using gravity extension.
7. Gravity extension lower the gears. Doors stay opened when downlocked.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: src/systems/systems/src/landing_gear/mod.rs [hunk 1]
```diff
@@ -265,55 +299,917 @@ impl LgciuGearExtension for LandingGearControlInterfaceUnit {
...
+        !(self.is_powered
+            && self.left_and_right_gear_extended(false)
+            && self.nose_gear_extended(false))
+    }
+}
+
+impl LgciuInterface for LandingGearControlInterfaceUnit {}
+
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum GearsSystemState {
```
Comment (BlueberryKing) [inline src/systems/systems/src/landing_gear/mod.rs:892]: There is a bit of a mix between references to a "gears system state" and "gear system state". I think choosing a consistent name makes sense.

```suggestion
pub enum GearSystemState {
```

File: src/failures/src/a320.ts [hunk 2]
```diff
@@ -15,6 +15,14 @@ export const A320Failure = Object.freeze({
...
     LeftPfdDisplay: 31000,
     RightPfdDisplay: 31001,
+    LgciuPowerSupply1: 32000,
+    LgciuPowerSupply2: 32001,
+    LgciuInternalError1: 32002,
+    LgciuInternalError2: 32003,
+    GearProxSensorDamageGearUplockLeft1: 32004,
+    GearProxSensorDamageDoorDownlockRight2: 32005,
+    GearProxSensorDamageGearUplockCenter1: 32006,
+    GearProxSensorDamageDoorUplockLeft2: 32007,
```
Comment (BlueberryKing) [inline src/failures/src/a320.ts:25]: It seems like there are some combinations missing here. Will you be adding those in a separate PR?

**Review by BlueberryKing — COMMENTED** [review]
Some initial comments. Note that I can't really review all the XML changes because I am not familiar with the format. However, I will try to finish my review of the Rust part soon. Please remind me if I have not finalized my review and you're ready to merge 😄.

File: src/systems/a320_systems/src/hydraulic/mod.rs [hunk 3]
```diff
@@ -4651,6 +5088,61 @@ impl SimulationElement for SpoilerComputer {
...
+
+    handle_angle: Angle,
+
+    is_extending_gear: bool,
+
+    is_turned: bool,
+}
+impl A320GravityExtension {
+    const INCREMENT_ANGLE_DEGREE_PER_SECOND: f64 = 120.;
+    const ALLOW_RETRACTION_DIRECTION: bool = false;
```
[the remark this answers is not in the document]
Comment (crocket63) [reply src/systems/a320_systems/src/hydraulic/mod.rs:5102]: You mean the retraction direction?  This one is because we have no animation for the emergency handle, so I prefer to disable ability to revert the emergency extension for now. It can be activated / option removed once we have a way to know where is the emergency handle.

File: src/systems/systems/src/landing_gear/mod.rs [hunk 4]
```diff
@@ -139,75 +103,145 @@ pub struct LandingGearControlInterfaceUnit {
...
+            GearWheel::CENTER => self.nose_gear_up_and_locked,
+            GearWheel::RIGHT => self.right_gear_up_and_locked,
+        };
+        let gear_downlocked = match wheel_id {
+            GearWheel::LEFT => self.left_gear_down_and_locked,
+            GearWheel::CENTER => self.nose_gear_down_and_locked,
+            GearWheel::RIGHT => self.right_gear_down_and_locked,
+        };
+
+        let in_transition = !(gear_downlocked ^ gear_uplocked);
```
[the remark this answers is not in the document]
Comment (crocket63) [reply src/systems/systems/src/landing_gear/mod.rs:212]: Locks info are coming from sensors, so typically anything can come from them even stupid stuff.

LGCIU implementation is pretty basic for now but I believe we'll have to add a lot of code that checks sensors if we can find how it's done in the real unit

File: src/failures/src/a320.ts (hunk 2 above)
Comment (crocket63) [reply src/failures/src/a320.ts:25]: There's already too much of them there and efb 2 display is getting crazy.... So yeah will add tons of them when efb GUI can handle this

**Comment by crocket63** [comment]
> Some initial comments. Note that I can't really review all the XML changes because I am not familiar with the format. However, I will try to finish my review of the Rust part soon. Please remind me if I have not finalized my review and you're ready to merge 😄. 

I don't know how xml work either lol

Thanks for the help though I know it's a big one, and it's far from being feature complete, but probably better to start as easy as possible and then add things

File: src/systems/a320_systems/src/hydraulic/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (BlueberryKing) [reply src/systems/a320_systems/src/hydraulic/mod.rs:5102]: Ah yeah, I meant the retraction direction. That makes sense then 👍.

**Review by BlueberryKing — COMMENTED** [review]
Two more comments.

File: src/systems/a320_systems/src/hydraulic/mod.rs [hunk 5]
```diff
@@ -1706,6 +2059,95 @@ impl HydraulicGeneratorControlUnit for A320Hydraulic {
     }
 }
 
+struct A320GearHydraulicController {
```
[the remark this answers is not in the document]
Comment (crocket63) [reply src/systems/a320_systems/src/hydraulic/mod.rs:2062]: It's an intermediate layer indeed, I'm trying here to group what's controlled by LGCIU outputs but also things that are not known from lgciu like the mechanical emergency extension system (vent valves / mechanical locks...).

[the remark this answers is not in the document]
Comment (tracernz) [reply src/systems/a320_systems/src/hydraulic/mod.rs:2062]: Keeping the mechanical/hydraulic simulation code separate to the LGCIU code that is modelled on the real LGCIU makes sense.

File: src/systems/a320_systems/src/hydraulic/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (crocket63) [reply src/systems/a320_systems/src/hydraulic/mod.rs]: Updated name to make this clear

**Review by BlueberryKing — APPROVED** [review]
Good to go from my side now 🥳!

**Comment by ShadowFrost3** [comment]
Quality Assurance Tester Report


Object of testing: #6893
Tier of Testing:1
Date: 04/27/2022

Testing Process: 
Normal operations:

Gear lever cannot be set to up position on ground whatever input event is used - checked
Gear can go up - checked/fail -- hardware variable issue -- decision needed
Gear can extend - checked/fail -- hardware variable issue -- decision needed

Gear cannot move above 264kts
No LGCIU fault ever apear in nominal conditions

Abnormal operations:

Gear cannot move without green hydraulics - checked 
Gear can move if using yellow epump plus PTU with all edp off. It should be slow however, and could cause LGCIU fault. - checked
LGCIU1 power failure cause an ecam warning and gear light indicator is non operative. Gear can still work normally if cycling gear lever - checked
LGCIU2 power failure cause an ecam warning and gear light indicator is still operative. Gear can still work normally if cycling gear lever - checked
Proximity sensor failures cause gear sequence to get stuck. Ecam fault will be trigger. After recycling gear lever gear operation should be normal. - checked 
Dual LGCIU power failure cause the gear to be unresponsive. Gear can be extended using gravity extension.
Gravity extension lower the gears. Doors stay opened when downlocked. - keybind only - checked - gear audio plays twice 

Note - it takes approximately 5 seconds for audio to be heard and gear to be released at 200kts - correct? Incorrect? 

Negatives: Hardware mismatch gear handle moves independently and only visually - should be tied to hardware - or completely disconnected not somewhere in between. Gear actions only triggered by hardware button. (Constant on in position if I remember correct) Its a two position switch. 

EFB failures page errors do not automatically resolve when errors are turned off - misclick potential 
EFB 3.0 needed - noted 

Testing Results: Not Passed - dependent on hardware bind decision 

Conclusions: Overall, the system seems well implemented and without the touched on items appears to work very well. There will need to be some discussions about the hardware item and whether that is allowable moving forward. But besides that, the system so far seems to be better than the problems I can throw at it.

**Comment by crocket63** [comment]
> 
> Note - it takes approximately 5 seconds for audio to be heard and gear to be released at 200kts - correct? Incorrect?
> 

As it's 5 seconds until it visually moves, this is ok as internally it's handling the door transition in the first 3 seconds.

> Negatives: Hardware mismatch gear handle moves independently and only visually - should be tied to hardware - or completely disconnected not somewhere in between. Gear actions only triggered by hardware button. (Constant on in position if I remember correct) Its a two position switch.
> 

For now, unless someone has some idea I'm unable to solve this. You have to chose if you use a button or cockpit clicks on lever

> 
> Hotspawn airstart gear position is broken
> 

Spawning gear up is not implemented in this PR.  I'll check if I can make this reliable by always starting gear down or with gear going up.

**Comment by aguther** [comment]
I guess the wheel page also needs an update.

**Comment by crocket63** [comment]
> I guess the wheel page also needs an update.

Definitly, separate PR for that task will be better though :)

File: src/systems/systems/src/landing_gear/mod.rs [hunk 6]
```diff
@@ -139,75 +103,145 @@ pub struct LandingGearControlInterfaceUnit {
...
+
+            nose_door_fully_opened: false,
+            right_door_fully_opened: false,
+            left_door_fully_opened: false,
+            nose_door_up_and_locked: false,
+            right_door_up_and_locked: false,
+            left_door_up_and_locked: false,
+
+            nose_gear_compressed_id: context.get_identifier(format!(
+                "LGCIU_{}_CENTER_GEAR_COMPRESSED",
```
Comment (tracernz) [inline src/systems/systems/src/landing_gear/mod.rs:144]: Please do not rename this. See ref FBW-32-01.

**Review by tracernz — DISMISSED** [review]

File: docs/a320-simvars.md [hunk 7]
```diff
@@ -2570,10 +2570,56 @@ In the variables below, {number} should be replaced with one item in the set: {
         - 1
         - 2
     - {gear}
-        - NOSE
+        - CENTER
```
Comment (tracernz) [inline docs/a320-simvars.md:2573]: Same here.

Comment (crocket63) [reply docs/a320-simvars.md:2573]: Getting lost here so before I start anything: Do I put everything NOSE ??  Or just do nothing?

**Comment by crocket63** [comment]
Ready to merge if @tracernz can check if changes are correct for nose naming :)

**Comment by Villason** [comment]
How does the gravity extension works? Clicked on the handle with all buttons but the gear does not goes down.

**Comment by crocket63** [comment]
Just click and hold, it simulates the handle doing all the 3 turns so can take couple of seconds
